### ISSUE TITLE
[FEATURE] Add request_model_include/exclude

### ISSUE THREAD

**Issue by skorokithakis** [tiangolo/fastapi#1357]
I have a model that include an ID field, which I return in responses. However, in requests (when POSTing new instances), I don't want the user to be able to specify the ID.

Similarly to how we can include/exclude fields in the response, we should be able to exclude fields in the request. Declaring more classes just to exclude specific fields is a bit cumbersome, and making the property optional is semantically wrong, since I don't want to let the user specify it at all.

**Comment by chris-allnutt** [comment]
You can create a special case CreateUser model, like what is used in the full stack example 

https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/schemas/user.py#L15

**Comment by skorokithakis** [comment]
Sure, but it feels dirty and unsemantic to have to create different models for each field you want to include, and is asymmetric with `response_model_exclude` existing.

**Comment by chris-allnutt** [comment]
Is there any reason you can't default the ID to None in the pydantic model?

**Comment by skorokithakis** [comment]
Yes, the ID is not optional, it must not be sent at all.

**Comment by Lunrtick** [comment]
As far as I can see, this would just require passing the exclude list straight to any calls to `models.to_dict` right? (as seen [here](https://pydantic-docs.helpmanual.io/usage/exporting_models/)).

**Comment by skorokithakis** [comment]
I think it's a bit more complicated than that, because the request is also for docs/requests, not just responses. I might have misunderstood what you mean, though, @Lunrtick, please correct me if so.

**Comment by Lunrtick** [comment]
I was talking about excluding it in the request, but you're right, it's definitely more complicated than just skipping the property in serialisation(mostly because of needing it to be documented in openapi). I was hoping to see how the response_model is translated to the openapi schema. I'd imagine that it wouldn't necessarily make sense to have a different Schema entry simply to exclude an ID?

**Comment by skorokithakis** [comment]
Yes, exactly, this is the rationale for the PR, so we can avoid having three different models just to exclude three fields. That's a bit clunky, whereas this solution is much more elegant and is symmetrical with the response include/exclude.

**Comment by RohitJones** [comment]
I think this also fits well with the OpenAPI 3.0 spec's [Read-Only and Write-Only Properties](https://swagger.io/docs/specification/data-models/data-types/). response_exclude maps well to writeOnly openapi property and a potential request_exclude would map nicely to the readOnly openapi property. 
However, I think that feature implementation might suit Pydantic better and have fastapi read off it.

**Comment by databasedav** [comment]
This doesn't address all of the concerns above but suffices for what I was looking for so might be helpful.

```python
from datetime import datetime
from typing import Iterable

from fastapi import FastAPI
from pydantic import BaseModel, create_model


app = FastAPI()

def require(model: BaseModel, *which: Iterable[str]):
    if not which:  # require all
        which = model.schema()['properties'].keys()
    return create_model(
        f"{model.__name__}__{'_'.join(which)}",
        __base__=model,
        **{w: (model.__fields__[w].outer_type_, ...) for w in which}
    )

class OmniUser(BaseModel):
    id: int = None
    email: str = None
    password: str = None
    created_at: datetime = None
    updated_at: datetime = None

@app.post('/create_user', response_model=require(OmniUser))
async def create_user(payload: require(OmniUser, 'email', 'password')):
    ...

```

This also carries over any `@validator`'s for the required fields.

**Comment by PrettyWood** [comment]
https://github.com/samuelcolvin/pydantic/pull/2231 may be helpful for this.
The remaining thing to do would be to change the generated schemas (add `readOnly: true`?)

**Comment by sadakmed** [comment]
> ```python
>         __base__=model,
> ```

`create_model` function inherits from `__base__` so  last argument in the function will filter the desired fields, but all the fields will be inherited from `model`, giving u the same inputted model, 

`__base__` should be `BaseModel`

**Comment by tdonovic** [comment]
A thought when I was trying to implement this was to create a base, request class, that excludes the id, created_at, updated_at fields, and then create another class that uses the first as its `__base__`. Is this an approach that would work? Still have the duplicate classes, but at least there is inheritance?

**Comment by mseimys** [comment]
IMHO It's perfectly fine that we need to define two different types for this use case. It's also fine to get those types by [combining](https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/schemas/user.py#L15) classes.

My reasoning: these objects are created and used in different context, it's just a coincidence that they match most of the properties, but imagine we add one more property like `client_id` which should register a user to some client, but it shouldn't be part of the `User` class. It will get confusing pretty fast. Also if you tried converting such "type" to other typed programming languages like TypeScript - you couldn't. You can't have two objects named the same, but in some context one has more properties than the other - no way to specify that.

For example, GraphQL even has dedicated "Input" types which kinda [try addressing the same issue you described](https://stackoverflow.com/questions/41743253/whats-the-point-of-input-type-in-graphql):
> Could you please explain why if input argument of mutation is object it should be input type? **I think much simpler just reuse type without providing id.**

The accepted answer describes some principles better than I did here :)

**Comment by InfinityMod** [comment]
> > ```python
> >         __base__=model,
> > ```
> 
> `create_model` function inherits from `__base__` so last argument in the function will filter the desired fields, but all the fields will be inherited from `model`, giving u the same inputted model,
> 
> `__base__` should be `BaseModel`
> 
> Edit: however the downside of this is the validators are also lost,
> 
> what I ended up doing is
> 
> ```
> In [98]: def delete(model: BaseModel, *fields_to_delete):
>     ...:     new_model = type('new_model', model.__bases__, dict(model.__dict__))
>     ...:     for field in fields_to_delete:
>     ...:         if field in new_model.__fields__:
>     ...:             del new_model.__fields__[field]
>     ...:     return new_model
> ```

I extended the function a little bit:

```
def filter(model: BaseModel, *which: Iterable[str], exclude=False):
    """Filters Pydantic Models and includes 
       (or excludes if set to True) the specified properties
       to abbreviate a model with a reduced field-set.
       All other properties are shared with the originated model."""
    if not which:  # require all
        which = model.schema()['properties'].keys()
    new_model = type(f"{model.__name__}__{'_'.join(which)}", model.__bases__,
                     dict(model.__dict__, __fields__=model.__fields__.copy()))
    if exclude:
        for field in which:
            if field in new_model.__fields__:
                del new_model.__fields__[field]
    else:
        for field in list(new_model.__fields__.keys()):
            if field not in which:
                del new_model.__fields__[field]
    return new_model

```

One small bug in the old one was, that __fields__ property of the new model pointed to the same __fields__ property of the old model, because dict(model.__dict__) only performed only a shallow copy. I solved it by adding an explicite copy of __fields__ to the new properties dict.

**Comment by Sgewux** [comment]
> This doesn't address all of the concerns above but suffices for what I was looking for so might be helpful.
> 
> ```python
> from datetime import datetime
> from typing import Iterable
> 
> from fastapi import FastAPI
> from pydantic import BaseModel, create_model
> 
> 
> app = FastAPI()
> 
> def require(model: BaseModel, *which: Iterable[str]):
>     if not which:  # require all
>         which = model.schema()['properties'].keys()
>     return create_model(
>         f"{model.__name__}__{'_'.join(which)}",
>         __base__=model,
>         **{w: (model.__fields__[w].outer_type_, ...) for w in which}
>     )
> 
> class OmniUser(BaseModel):
>     id: int = None
>     email: str = None
>     password: str = None
>     created_at: datetime = None
>     updated_at: datetime = None
> 
> @app.post('/create_user', response_model=require(OmniUser))
> async def create_user(payload: require(OmniUser, 'email', 'password')):
>     ...
> ```
> 
> This also carries over any `@validator`'s for the required fields.
> 
> **edit:** change `model.__fields__[w].outer_type_` to `typing.Optional[model.__fields__[w].outer_type_]` to require the fields but allow `None`

Based in this code i created a classmethod for one of my models to exclude some fields of it (and avoid definig a buch of models)

```python    @classmethod
    def exclude(cls, *fields: Iterable[str]) -> BaseModel:

        field_definition = {}
        model_fields = cls.__fields__

        for field in model_fields.keys():
            if field not in fields:
                model_field = model_fields[field]
                if model_field.required:
                    field_definition[field] = (model_field.outer_type_, ...)
                else:
                    field_definition[field] = (model_field.outer_type_, model_field.default)
                    

        new_model = create_model(
            __model_name = f'{cls.__name__}_without_{"_".join(fields)}',
            __base__=BaseModel,
            **field_definition
        )

        return new_model```

**Comment by loay106** [comment]
is this supported with the new version of Fast API?
this enhancement can be very useful IMHO.
